design/lvm/mountable_snapshot.md

```diff
@@ -0,0 +1,191 @@
+      f:metadata:
+        f:annotations:
+          f:pv.kubernetes.io/bind-completed: {}
+          f:pv.kubernetes.io/bound-by-controller: {}
+          f:volume.beta.kubernetes.io/storage-provisioner: {}
+          f:volume.kubernetes.io/storage-provisioner: {}
+      f:spec:
+        f:volumeName: {}
+    manager: kube-controller-manager
+    operation: Update
+    time: "2022-03-31T10:37:06Z"
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:status:
+        f:accessModes: {}
+        f:capacity:
+          .: {}
+          f:storage: {}
+        f:phase: {}
+    manager: kube-controller-manager
+    operation: Update
+    subresource: status
+    time: "2022-03-31T10:37:06Z"
+  - apiVersion: v1
+    fieldsType: FieldsV1
+    fieldsV1:
+      f:metadata:
+        f:annotations:
+          .: {}
+          f:kubectl.kubernetes.io/last-applied-configuration: {}
+      f:spec:
+        f:accessModes: {}
+        f:dataSource: {}
+        f:resources:
+          f:requests:
+            .: {}
+            f:storage: {}
+        f:storageClassName: {}
+        f:volumeMode: {}
+    manager: kubectl
+    operation: Update
+    time: "2022-03-31T10:37:06Z"
+  name: csi-lvmpvc-snap
+  namespace: default
+  resourceVersion: "10994"
+  uid: 0217ba5b-0ba2-4e28-af2a-acb97c55e8b1
+spec:
+  accessModes:
+  - ReadWriteOnce
+  dataSource:
+    apiGroup: snapshot.storage.k8s.io
+    kind: VolumeSnapshot
+    name: lvmpv-snap
+  resources:
+    requests:
+      storage: 10Gi
+  storageClassName: openebs-lvmpv
+  volumeMode: Filesystem
+  volumeName: pvc-0217ba5b-0ba2-4e28-af2a-acb97c55e8b1
+status:
+  accessModes:
+  - ReadWriteOnce
+  capacity:
+    storage: 10Gi
+  phase: Bound
+```
+
+## Implementation Details
+
+In order to implement this only existing code needs to be changed, there is no need for an entirely new control flow.
+The changes below are the primary modules that need to be modified, however, there are some additional utility modules that will need to be modified to support these changes. Most likely the following files will need some additions to support snapshots: [kubernetes.go](../../pkg/builder/volbuilder/kubernetes.go) (in order to implement a `Kubeclient.GetSnapshot` function), [lvm_util.go](../../pkg/lvm/lvm_util.go) (in order to create a function that changes the snapshot write access) and [mount.go](../../pkg/lvm/mount.go) (in order to support mounting/unmounting of snapshots).
+
+- In the [controller.go](../../pkg/driver/controller.go) the code path in the `CreateVolume` function for the `controller` type that occurs when contentSource.GetSnapshot() is not `nil` needs to be implemented. When this path is triggered it should return the correct `volName`, `topology` and `cntx`.  
+- In the [agent.go](../../pkg/driver/agent.go) the `NodePublishVolume` function for the `node` type needs to be changed such that it checks whether the `volName` is a snapshot and if so mount the snapshot to the specified location.
```

Comment:
Why do we need to know whether the `volName` is a snapshot? We are creating a volume out of the snapshot and then simply we want to mount that newly created volume. Am I interpreting it right or am I missing something?

Reply:
I figured we'd need to call different functions so that we end up with an `*apis.LVMSnapshot` object instead of a `*apis.LVMVolume` object, but that might not even be needed. I suppose that's mainly an implementation detail and this should simply say that it mount the snapshot.